tools/generate_config.py

```diff
@@ -0,0 +1,326 @@
+
+# Validates a particular enclave_cfg_path against ENCLAVE_CFG_SCHEMA_PATH
+def validate(args):
+    if jsonschema is None:
+        raise RuntimeError('validate requires the "jsonschema" Python package')
+    schema_path = ENCLAVE_CFG_SCHEMA_PATH
+    print('Validating {args.file}')
+    with open(args.file) as f:
+      cfg = json.load(f)
+    with open(schema_path) as f:
+      cfg_schema = json.load(f)
+    jsonschema.validate(instance=cfg, schema=cfg_schema)
+    print('No errors found.')
+
+def post_type(jtype):
+  if 'type' in jtype:
+    jtt = jtype['type']
+    if 'maxLength' in jtype:
+      if jtt == 'string':
+        return '[' + str(jtype['maxLength']+1) + ']'
+      elif jtt == 'array':
+        return '[' + str(jtype['maxLength']) + ']'
+    return ''
+  else:
+    return ''
+
+def pre_type(jtype):
+  if 'type' in jtype:
+    jtt = jtype['type']
+    if jtt == 'array' or (isinstance(jtt, list) and 'array' in jtt and 'null' in jtt):
+      items = jtype['items']
+      item_type = pre_type(items)
+      if 'maxLength' in jtype:
+        return item_type
+      else:
+        return item_type + '*'
+    elif jtt == 'boolean': return 'bool'
+    elif jtt == 'none': return 'null'
+    elif jtt == 'string' or (isinstance(jtt, list) and 'string' in jtt and 'null' in jtt):
+      return 'char' if 'maxLength' in jtype else 'char*'
+    else:
+        print('unhandled json type: %s' % jtype)
+  elif '$ref' in jtype:
+    rtype = jtype['$ref'][jtype['$ref'].rfind('/')+1:]
+    if rtype.startswith('safe_'):
+      return rtype[5:]
+    else:
+      return rtype
+  else:
+    print('unknown json type: %s' % jtype)
+    return jtype
+
+def need_size_var(jtype):
+  if 'type' not in jtype:
+    return False
+  jtt = jtype['type']
+  return (jtt == 'array' and 'maxLength' not in jtype) or (isinstance(jtt, list) and 'array' in jtt and 'null' in jtt)
+
+num_settings = 0
+
+def generate_header(schema_file_name, root, args):
+  global num_settings
+
+  with open(str(args.header), "w") as header:
+    header.write('#ifndef SGXLKL_ENCLAVE_CONFIG_GEN_H\n');
+    header.write('#define SGXLKL_ENCLAVE_CONFIG_GEN_H\n');
+
+    header.write('\n/* Automatically generated from %s; do not modify. */\n\n' % schema_file_name);
+
+    header.write('#include <inttypes.h>\n');
+    header.write('#include <stdbool.h>\n');
+    header.write('#include <stddef.h>\n');
+    header.write('#include <elf.h>\n\n');
+
+    header.write('#define SGXLKL_ENCLAVE_CONFIG_VERSION 1UL\n\n');
+
+    num_settings = 0
+    for typename, typedef in root['definitions'].items():
+      if (typename.startswith('sgxlkl_')):
+        if 'enum' in typedef:
+          names = typedef['c_enum'] if 'c_enum' in typedef else typedef['enum']
+          header.write('typedef enum\n{\n')
+          i = 0
+          num_vals = len(names)
+          for value in names:
+            header.write('    %s = %d' % (value.upper(), i))
+            if i < num_vals - 1: header.write(',')
```

Comment:
Please always use a new line, just for consistency.

Reply:
Not sure what you're referring to. Those two `header.write` correctly output one line of text. The newline follows two lines later.

Reply:
I just meant Python code style, don't add a command directly after `:` of an if statement. It's easy to overlook that.

Reply:
Aah, I see, sure I can change that.